zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -0,0 +1,1156 @@
+            case OpCode.setWatches:
+            case OpCode.setWatches2:
+            case OpCode.checkWatches:
+            case OpCode.removeWatches:
+            case OpCode.getEphemerals:
+            case OpCode.multiRead:
+            case OpCode.addWatch:
+            case OpCode.whoAmI:
+                zks.sessionTracker.checkSession(request.sessionId, request.getOwner());
+                break;
+            default:
+                LOG.warn("unknown type {}", request.type);
+                break;
+            }
+        } catch (KeeperException e) {
+            if (request.getHdr() != null) {
+                request.getHdr().setType(OpCode.error);
+                request.setTxn(new ErrorTxn(e.code().intValue()));
+            }
+
+            if (e.code().intValue() > Code.APIERROR.intValue()) {
+                LOG.info(
+                    "Got user-level KeeperException when processing {} Error Path:{} Error:{}",
+                    request.toString(),
+                    e.getPath(),
+                    e.getMessage());
+            }
+            request.setException(e);
+        } catch (Exception e) {
+            // log at error level as we are returning a marshalling
+            // error to the user
+            LOG.error("Failed to process {}", request, e);
+
+            StringBuilder sb = new StringBuilder();
+            ByteBuffer bb = request.request;
+            if (bb != null) {
+                bb.rewind();
+                while (bb.hasRemaining()) {
+                    sb.append(Integer.toHexString(bb.get() & 0xff));
+                }
+            } else {
+                sb.append("request buffer is null");
+            }
+
+            LOG.error("Dumping request buffer: 0x{}", sb.toString());
+            if (request.getHdr() != null) {
+                request.getHdr().setType(OpCode.error);
+                request.setTxn(new ErrorTxn(Code.MARSHALLINGERROR.intValue()));
+            }
+        }
+    }
+
+    private static List<ACL> removeDuplicates(final List<ACL> acls) {
+        if (acls == null || acls.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        // This would be done better with a Set but ACL hashcode/equals do not
+        // allow for null values
+        final ArrayList<ACL> retval = new ArrayList<>(acls.size());
+        for (final ACL acl : acls) {
+            if (!retval.contains(acl)) {
+                retval.add(acl);
+            }
+        }
+        return retval;
+    }
+
+    private void validateCreateRequest(String path, CreateMode createMode, Request request, long ttl) throws KeeperException {
+        if (createMode.isTTL() && !EphemeralType.extendedEphemeralTypesEnabled()) {
+            throw new KeeperException.UnimplementedException();
+        }
+        try {
+            EphemeralType.validateTTL(createMode, ttl);
+        } catch (IllegalArgumentException e) {
+            throw new BadArgumentsException(path);
+        }
```

Comment:
### ❌ New issue: Complex Method
validateCreateRequest has a cyclomatic complexity of 9, threshold = 9